Product/Production/Services/PatientDiscoveryCore/src/main/java/gov/hhs/fha/nhinc/patientdiscovery/audit/transform/AbstractPatientDiscoveryAuditTransforms.java

```diff
@@ -0,0 +1,216 @@
+            PatientDiscoveryAuditTransformsConstants.PARTICIPANT_QUERYPARAMS_OBJ_TYPE_CODE_ROLE,
+            PatientDiscoveryAuditTransformsConstants.PARTICIPANT_QUERYPARAMS_OBJ_ID_TYPE_CODE,
+            PatientDiscoveryAuditTransformsConstants.PARTICIPANT_QUERYPARAMS_OBJ_ID_TYPE_CODE_SYSTEM,
+            PatientDiscoveryAuditTransformsConstants.PARTICIPANT_QUERYPARAMS_OBJ_ID_TYPE_DISPLAY_NAME);
+        participantObject.setParticipantObjectID(participantObjectId);
+        return participantObject;
+    }
+
+    protected Marshaller getMarshaller() throws JAXBException {
+        return new JAXBContextHandler().getJAXBContext(JAXB_HL7_CONTEXT_NAME).createMarshaller();
+    }
+
+    @Override
+    protected String getServiceEventIdCodeRequestor() {
+        return PatientDiscoveryAuditTransformsConstants.EVENT_ID_CODE;
+    }
+
+    @Override
+    protected String getServiceEventIdCodeResponder() {
+        return PatientDiscoveryAuditTransformsConstants.EVENT_ID_CODE;
+    }
+
+    @Override
+    protected String getServiceEventCodeSystem() {
+        return PatientDiscoveryAuditTransformsConstants.EVENT_CODE_SYSTEM;
+    }
+
+    @Override
+    protected String getServiceEventDisplayRequestor() {
+        return PatientDiscoveryAuditTransformsConstants.EVENT_CODE_DISPLAY_REQUESTOR;
+    }
+
+    @Override
+    protected String getServiceEventDisplayResponder() {
+        return PatientDiscoveryAuditTransformsConstants.EVENT_CODE_DISPLAY_RESPONDER;
+    }
+
+    @Override
+    protected String getServiceEventTypeCode() {
+        return PatientDiscoveryAuditTransformsConstants.EVENT_TYPE_CODE;
+    }
+
+    @Override
+    protected String getServiceEventTypeCodeSystem() {
+        return PatientDiscoveryAuditTransformsConstants.EVENT_TYPE_CODE_SYSTEM;
+    }
+
+    @Override
+    protected String getServiceEventTypeCodeDisplayName() {
+        return PatientDiscoveryAuditTransformsConstants.EVENT_TYPE_CODE_DISPLAY_NAME;
+    }
+
+    @Override
+    protected String getServiceEventActionCodeRequestor() {
+        return PatientDiscoveryAuditTransformsConstants.EVENT_ACTION_CODE_REQUESTOR;
+    }
+
+    @Override
+    protected String getServiceEventActionCodeResponder() {
+        return PatientDiscoveryAuditTransformsConstants.EVENT_ACTION_CODE_RESPONDER;
+    }
+
+    protected AuditMessageType getPatientParticipantObjectIdentificationForResponse(PRPAIN201306UV02 response,
+        AuditMessageType auditMsg) {
+
+        List<II> oII = PRPAIN201306UV02Parser.getPatientIds(response);
+        if (oII != null && oII.size() > 0) {
+            for (II entry : oII) {
+                if (entry != null && entry.getRoot() != null && entry.getExtension() != null
+                    && !entry.getRoot().isEmpty() && !entry.getExtension().isEmpty()) {
+                    createPatientParticipantObjectIdentification(auditMsg, entry.getRoot(), entry.getExtension());
```

Comment:
Also: Do we create a Patient Participant Object Identification for every non-null, non-empty Root / Extension combination, or should there be a break here?